go/src/koding/klient/machine/index/index.go

```diff
@@ -135,22 +140,41 @@ func NewIndexFiles(root string) (*Index, error) {
 			return nil
 		}
 
-		entry, err := NewEntryFile(root, path, info)
-		if err != nil {
-			return nil
-		}
-
-		idx.entries[entry.Name] = entry
+		wg.Add(1)
+		idx.addEntry(&wg, root, path, info)
 		return nil
 	}
 
 	if err := filepath.Walk(root, walkFn); err != nil {
 		return nil, err
 	}
 
+	wg.Wait()
 	return idx, nil
 }
 
+// addEntry asynchronously adds new entry to index. This function has limits
+// to a number of started go-routines. Errors are ignored.
+func (idx *Index) addEntry(wg *sync.WaitGroup, root, path string, info os.FileInfo) {
+	idx.limitC <- struct{}{}
+
+	go func() {
```

Comment:
could we change the structure here a bit with goroutine worker pool pattern? Instead of creating goroutines per file operation we can limit it by `2*runtime.NumCPU()` 

just to clarify what i mean by [worker pool](https://gobyexample.com/worker-pools)